### PR TITLE
LibCore: Refactor in Core::Stream

### PR DESCRIPTION
**Pull request by LucasChollet**
This is preliminary work to ease the introduction of #16437.

cc @timschumi

### CODE REVIEW

**Comment by timschumi** [comment]
Probably won't have time until Monday to look at this in detail, so it's just a few preliminary things for now:

- First commit definitely needs a better title.
- First commit might also need some more splitting up, preferably in "move buffering into a new function" and "use new function to buffer in `can_read_line`"

To me, this PR as a whole (especially the two last commits, and the first commit kind of as well) doesn't seem all that related to making the `CircularBuffer` switch easier (or at least I don't see it, since the code size and used APIs don't appear to change significantly).

We could (and probably should) move #16437 forward first, considering that it's gotten more review until now and generally seems like a less intrusive change, as it is (in its core) "just" a new general-purpose data structure and some API replacements. If I can find a bit of time tomorrow morning I could also try and do some decoupling.

**Comment by LucasChollet** [comment]
I will do it the other way around, so it's a draft until #16437 is merged

**Comment by AtkinsSJ** [comment]
Oh hey, the PR this was waiting on got merged. :eyes:

**Comment by LucasChollet** [comment]
I'm currently working on it :sweat_smile:

File: Userland/Libraries/LibCore/Stream.h
Comment (timschumi) [reply Userland/Libraries/LibCore/Stream.h:677]: Regarding this change, doing a `find_and_populate_until_any_of` won't change the size of the `buffer` anyways (as that denotes the passed `Span`, not the internal buffer), so we should probably be doing the `ENOBUFS` first indeed.

**Review by timschumi — CHANGES_REQUESTED** [review]

File: AK/CircularBuffer.h
Comment (LucasChollet) [reply AK/CircularBuffer.h:49]: Yeah, going with the `Optional` is IMO better for consistency.

File: AK/CircularBuffer.cpp
Comment (LucasChollet) [reply AK/CircularBuffer.cpp]: I understand your point and replaced it with `from`, not sure that it's good tho

Comment (timschumi) [reply AK/CircularBuffer.cpp]: I feel like it's at least much clearer than `after`. Comment is resolved in my opinion.

File: Userland/Libraries/LibCore/Stream.h
Comment (timschumi) [inline Userland/Libraries/LibCore/Stream.h:680]: This looks like only the updated commit message made it in, but not the actual change.

File: Userland/Libraries/LibCore/Stream.h
Comment (timschumi) [inline Userland/Libraries/LibCore/Stream.h]: Note that the original intention behind this `FIXME` was that we already searched through most of the buffer on a previous iteration and now only need to search the bytes that were freshly appended by the former `populate_read_buffer` (plus a bit of overlap for long patterns). I guess this works fine as well, but we need to make sure that we correctly handle overlapping patterns and that the order of the patterns doesn't have an effect on the result.

File: Userland/Libraries/LibCore/Stream.h
Comment (timschumi) [inline Userland/Libraries/LibCore/Stream.h]: Missing space after `{},`, unchecked by the linter because it's removed in a future commit.

**Review by timschumi — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibCore/Stream.h
Comment (LucasChollet) [reply Userland/Libraries/LibCore/Stream.h:680]: Oops, I also removed another occurrence in `BufferedHelper::read()`

File: Userland/Libraries/LibCore/Stream.h
Comment (LucasChollet) [reply Userland/Libraries/LibCore/Stream.h]: I don't understand what you are expecting me to do.
I think that solves the overlapping issue.
```cpp
auto const corrected_minimum_offset = *longest_candidate > min_offset ? 0 : min_offset - *longest_candidate;
[...]
for (auto& candidate : candidates) {
    auto const result = m_buffer.offset_of(candidate, corrected_minimum_offset, *max_offset);
```
And AFAICT, the code that matters for pattern's order is directly copied from the old implementation.

Comment (timschumi) [reply Userland/Libraries/LibCore/Stream.h]: I just wanted to bring to your attention that what you did here is beyond the scope of what I originally intended with that FIXME. As long as everything works as intended there is no need for you to do anything, it just seemed like a slightly larger attack surface for edge-cases that was worth mentioning.

Comment (LucasChollet) [reply Userland/Libraries/LibCore/Stream.h]: Oh okay, for sure I didn't understand then.

File: Userland/Libraries/LibCore/Stream.h
Comment (timschumi) [inline Userland/Libraries/LibCore/Stream.h:678]: Considering that `find_and_populate_until_any_of` with `max_offset` being 0 basically would be a no-op anyways, I think it would be more logical to keep the empty buffer check where it is and just change the response (as with the other check).

```cpp
[...]

if (buffer.is_empty())
    return buffer;

auto const candidate = TRY(find_and_populate_until_any_of(candidates, buffer.size()));

[...]
```

**Review by timschumi — APPROVED** [review]
One last small nitpick, but feel free to ignore. Otherwise, this looks good to me.

File: Userland/Libraries/LibCore/Stream.h
Comment (timschumi) [inline Userland/Libraries/LibCore/Stream.h:710]: Also, a very minor nitpick, but this feels quite verbose. How about the following?

```suggestion
    struct Match {
        size_t offset {};
        size_t size {};
    };
```

**Review by timschumi — APPROVED** [review]

**Review by ADKaster — APPROVED** [review]

**Pull request merged**
